File: src/data/usageLogs.js
import { differenceInCalendarDays } from 'date-fns'
import { activeEquipment } from './equipment.js'
import { genDailyLogs } from '../lib/genLogs.js'
import { SIM_TODAY } from '../lib/clock.js'

// Two units are seeded to misbehave so the geofence and fuel rules have
// something real to catch on first load. Everything else logs normally.
const SEEDED_ANOMALIES = {
  'EQX-2025': { driftKm: 18.5 }, // wanders 18.5km to Tambaram Railway Yard
  'EQX-2003': { driftKm: 18.5 }, // wanders 18.5km to Tambaram Railway Yard
  'EQX-2016': { fuelFactor: 1.8 }, // burning ~80% more fuel per engine hour than expected
}

// Daily usage logs for every active rental, from check-in through "today".
// This is what the utilization charts, reports, and the anomaly/recommendation
// rules all read from — completed rentals already carry their aggregate
// avgEngineHoursPerDay / avgIdleHoursPerDay directly on the equipment record.
export const usageLogs = activeEquipment.flatMap((eq) => {
  const days = Math.max(1, differenceInCalendarDays(SIM_TODAY, new Date(eq.checkIn)) + 1)
  return genDailyLogs({
    equipmentId: eq.id,
    operatorId: eq.operatorId,
    siteId: eq.siteId,
    startDate: eq.checkIn,
    days,
    baseEngine: eq.avgEngineHoursPerDay,
    baseIdle: eq.avgIdleHoursPerDay,
    ...(SEEDED_ANOMALIES[eq.id] ?? {}),
  })
})

export function logsFor(equipmentId) {
  return usageLogs.filter((l) => l.equipmentId === equipmentId)
}

/** Most recent log per equipment id — what the anomaly rules evaluate. */
export function latestLogByEquipment(logs) {
  const latest = {}
  for (const log of logs) {
    if (!latest[log.equipmentId] || log.date > latest[log.equipmentId].date) {
      latest[log.equipmentId] = log
    }
  }
  return latest
}

File: backend/app/seed_data/usage_logs.py
"""Port of src/data/usageLogs.js — daily usage logs for every active rental,
from check-in through the simulated "today". Two units are seeded to
misbehave (matching the original JS fixture exactly) so the geofence/fuel
alert rules have something real to catch on first load.
"""

from datetime import date

from .equipment import SIM_TODAY
from .generate_logs import gen_daily_logs
from .sites import SITE_BY_ID

SEEDED_ANOMALIES = {
    "EQX-2003": {"drift_km": 6.5},   # wanders outside Hillside Quarry's radius
    "EQX-2016": {"fuel_factor": 1.8},  # burns ~80% more fuel per engine hour than expected
}


def build_usage_logs(equipment: list[dict]) -> list[dict]:
    logs: list[dict] = []
    for eq in equipment:
        if eq["status"] != "active":
            continue
        check_in = date.fromisoformat(eq["check_in"])
        days = max(1, (SIM_TODAY - check_in).days + 1)
        anomaly = SEEDED_ANOMALIES.get(eq["id"], {})
        logs.extend(
            gen_daily_logs(
                equipment_id=eq["id"],
                operator_id=eq.get("operator_id"),
                site_id=eq.get("site_id"),
                start_date=eq["check_in"],
                days=days,
                base_engine=eq["avg_engine_hours_per_day"],
                base_idle=eq["avg_idle_hours_per_day"],
                site_by_id=SITE_BY_ID,
                **anomaly,
            )
        )
    return logs
